### PR TITLE
Adding support for deal prefixes

### PR DESCRIPTION
**Pull request by camrice**
Opening PR to get an early start on review comments. We have a couple questions remaining on our end. One major change that may need to be added is implementing the `pmp` object for targeting deals in the impressions. I will update this PR with any action items regarding this.

### CODE REVIEW

**Comment by camrice** [comment]
Relevant issue for this PR with general documentation: #1174

**Comment by camrice** [comment]
Update to my initial comment:

The `pmp` object is not required for supporting deals, so this change is ready for review. There aren't any other major additions needed at this time.

File: exchange/exchange.go [hunk 1]
```diff
@@ -167,12 +169,95 @@ func (e *exchange) HoldAuction(ctx context.Context, bidRequest *openrtb.BidReque
...
+
+// Updates targeting keys with deal prefixes if minimum deal tier exceeded
+func applyDealSupport(bidRequest *openrtb.BidRequest, auc *auction) []error {
+	errs := []error{}
+	impDealMap := getDealTiers(bidRequest)
+
+	for impID, topBidsPerImp := range auc.winningBidsByBidder {
+		impDeal := impDealMap[impID].DealInfo
+		for bidderName, topBidPerBidder := range topBidsPerImp {
+			bidderString := bidderName.String()
```
Comment (SyntaxNode) [inline exchange/exchange.go]: Nitpick: Including the type of the variable in the name is a bit of a 'smell' IMHO. Can we simply name it `bidder`?

File: exchange/exchange.go [hunk 2]
```diff
@@ -167,12 +169,95 @@ func (e *exchange) HoldAuction(ctx context.Context, bidRequest *openrtb.BidReque
...
+					errs = append(errs, fmt.Errorf("dealTier configuration invalid for bidder '%s', imp ID '%s'", bidderString, impID))
+				}
+			}
+		}
+	}
+
+	return errs
+}
+
+// Creates map of impression to bidder deal tier configuration
```
Comment (SyntaxNode) [inline exchange/exchange.go]: Should this follow the go doc convention of `getDealTiers creates a map of...`?

File: exchange/exchange.go [hunk 3]
```diff
@@ -167,12 +169,95 @@ func (e *exchange) HoldAuction(ctx context.Context, bidRequest *openrtb.BidReque
...
+		impDealMap[imp.ID] = &bidderDealTier
+	}
+
+	return impDealMap
+}
+
+func validateDealTier(impDeal *DealTier) bool {
+	if impDeal.Info != nil {
+		// Remove whitespace from prefix before checking if it can be used
+		impDeal.Info.Prefix = regexp.MustCompile(`\s+`).ReplaceAllString(impDeal.Info.Prefix, "")
```
Comment (SyntaxNode) [inline exchange/exchange.go]: Is it common for a prefix to include whitespace? aka... why is this needed?

File: exchange/exchange.go [hunk 4]
```diff
@@ -167,12 +169,95 @@ func (e *exchange) HoldAuction(ctx context.Context, bidRequest *openrtb.BidReque
...
+		return len(impDeal.Info.Prefix) > 0 && impDeal.Info.MinDealTier > 0
+	}
+	return false
+}
+
+func updateCatDur(bid *pbsOrtbBid, dealTierInfo *DealTierInfo) {
```
Comment (SyntaxNode) [inline exchange/exchange.go]: This is likely industry knowledge I'm lacking, but what is a 'CatDur'?

File: exchange/exchange.go [hunk 5]
```diff
@@ -167,12 +169,95 @@ func (e *exchange) HoldAuction(ctx context.Context, bidRequest *openrtb.BidReque
...
+	if bid.dealPriority >= dealTierInfo.MinDealTier {
+		newPb := fmt.Sprintf("%s%d_", dealTierInfo.Prefix, bid.dealPriority)
```
Comment (SyntaxNode) [inline exchange/exchange.go]: Could `newPb` have a more descriptive name?

File: adapters/appnexus/appnexus.go [hunk 6]
```diff
@@ -87,6 +87,7 @@ type appnexusBidExtAppnexus struct {
 	BrandId       int                    `json:"brand_id"`
 	BrandCategory int                    `json:"brand_category_id"`
 	CreativeInfo  appnexusBidExtCreative `json:"creative_info"`
+	DealPriority  int                    `json:"deal_priority"`
```
Comment (SyntaxNode) [inline adapters/appnexus/appnexus.go]: To be clear, is this just modeling the response from AppNexus? I'm really asking if an int is an appropriate type or if it should be an int pointer or if an omit empty attribute should be added to the field.

File: exchange/exchange.go [hunk 7]
```diff
@@ -167,12 +169,95 @@ func (e *exchange) HoldAuction(ctx context.Context, bidRequest *openrtb.BidReque
 			}
 			targData.setTargeting(auc, bidRequest.App != nil, bidCategory)
 		}
+
+		if requestExt.Prebid.SupportDeals {
+			dealErrs := applyDealSupport(bidRequest, auc)
+			if len(dealErrs) > 0 {
```
Comment (SyntaxNode) [inline exchange/exchange.go]: Is the length check needed? Go's builtin methods are kind to nil or empty collections, so `errs = append(errs, dealErrs...)` might work just fine.

File: exchange/exchange_test.go [hunk 8]
```diff
@@ -1325,6 +1325,267 @@ func TestUpdateRejections(t *testing.T) {
...
+	bidderName2 := openrtb_ext.BidderName("validmultiple2")
+	bidderName3 := openrtb_ext.BidderName("minnotmet")
+	bidderName4 := openrtb_ext.BidderName("invalidconfig")
+	bidderName5 := openrtb_ext.BidderName("nodeal")
+	bidderName6 := openrtb_ext.BidderName("invalidconfig2")
+
+	bidRequest := &openrtb.BidRequest{
+		ID: "some-request-id",
+		Imp: []openrtb.Imp{
+			{
```
Comment (SyntaxNode) [inline exchange/exchange_test.go]: Why does this test need 6 imps? If each one is intended to be a unique test case, consider breaking it down using table driven tests. Same feedback for the other tests in this file.

File: exchange/exchange.go [hunk 9]
```diff
@@ -167,12 +169,95 @@ func (e *exchange) HoldAuction(ctx context.Context, bidRequest *openrtb.BidReque
...
+		}
+
+		impDealMap[imp.ID] = &bidderDealTier
+	}
+
+	return impDealMap
+}
+
+func validateDealTier(impDeal *DealTier) bool {
+	if impDeal.Info != nil {
```
Comment (SyntaxNode) [inline exchange/exchange.go]: Consider switching the logic and including the `return false` up here for a quick exit, leaving the rest of the code not indented down a level.

File: exchange/exchange.go [hunk 10]
```diff
@@ -167,12 +169,95 @@ func (e *exchange) HoldAuction(ctx context.Context, bidRequest *openrtb.BidReque
...
+	return errs
+}
+
+// Creates map of impression to bidder deal tier configuration
+func getDealTiers(bidRequest *openrtb.BidRequest) map[string]*BidderDealTier {
+	impDealMap := make(map[string]*BidderDealTier)
+
+	for _, imp := range bidRequest.Imp {
+		var bidderDealTier BidderDealTier
+		err := json.Unmarshal(imp.Ext, &bidderDealTier.DealInfo)
```
Comment (SyntaxNode) [inline exchange/exchange.go]: Would this information have already been parsed previously somewhere else in code? Is there any way to avoid this unmarshal call?

File: exchange/exchange.go (hunk 3 above)
Comment (camrice) [reply exchange/exchange.go]: The prefixes shouldn't contain whitespace, but currently the UI for the prebid configs doesn't have any validation or blocking for this case. This is used to just guard against those cases until the UI is able to validate it. Would it be better to drop the prefix modification if it does have whitespace?

File: exchange/exchange.go (hunk 1 above)
Comment (camrice) [reply exchange/exchange.go]: Ah yeah the variable name is unnecessary, updated to `bidder`

File: exchange/exchange.go (hunk 2 above)
Comment (camrice) [reply exchange/exchange.go]: Updated comment to follow doc convention

File: exchange/exchange.go (hunk 4 above)
Comment (camrice) [reply exchange/exchange.go]: 'CatDur' is a shortening of the key value `hb_pb_cat_dur` but I don't think it is a standard shortening. I will update this function to be `updateHbPbCatDur` so that it is more specific.

File: exchange/exchange.go (hunk 3 above)
Comment (SyntaxNode) [reply exchange/exchange.go]: Gotcha. I'm asking because the use of the regex raises a performance red flag and was going to see if a string trim method could be used instead.

File: exchange/exchange.go (hunk 4 above)
Comment (SyntaxNode) [reply exchange/exchange.go]: Thank you. That makes sense.

File: exchange/exchange.go (hunk 3 above)
Comment (camrice) [reply exchange/exchange.go]: Understood! Ideally this kind of check wouldn't be needed. I mainly used the regex to remove newlines and tabs, though that seems excessive now. I could use a string replace or trim instead to avoid this.

File: exchange/exchange.go (hunk 5 above)
Comment (camrice) [reply exchange/exchange.go]: Ah yes this isn't a price bucket anymore. Updated to be `prefixTier`.

File: exchange/exchange.go (hunk 7 above)
Comment (camrice) [reply exchange/exchange.go]: Length check is indeed not needed, removed.

File: exchange/exchange.go (hunk 9 above)
Comment (camrice) [reply exchange/exchange.go]: Flipped the logic here so that the nil check will lead to false first

File: exchange/exchange_test.go (hunk 8 above)
Comment (camrice) [reply exchange/exchange_test.go]: Refactored all added unit tests to use the table driven test approach. Also removed a couple test cases that were redundant.

File: exchange/exchange.go (hunk 10 above)
Comment (camrice) [reply exchange/exchange.go]: I don't think the impression ext is unmarshalled elsewhere, at least not with the deal tier info. It seems the extension is used in the auction endpoint, but the map created there is bidder to raw message.

File: adapters/appnexus/appnexus.go (hunk 6 above)
Comment (camrice) [reply adapters/appnexus/appnexus.go]: I think here the int is the appropriate type since it will initialize to 0 from the `TypedBid` struct. A value of 0 here would be considered a bid without a deal.

File: exchange/exchange.go (hunk 3 above)
Comment (camrice) [reply exchange/exchange.go]: Updated to use `strings.ReplaceAll` instead of regex

File: adapters/appnexus/appnexus.go (hunk 6 above)
Comment (SyntaxNode) [reply adapters/appnexus/appnexus.go]: Ok, cool.

File: exchange/exchange.go [hunk 11]
```diff
@@ -167,12 +168,93 @@ func (e *exchange) HoldAuction(ctx context.Context, bidRequest *openrtb.BidReque
...
+type DealTier struct {
+	Info *DealTierInfo `json:"dealTier,omitempty"`
+}
+
+type BidderDealTier struct {
+	DealInfo map[string]*DealTier
+}
+
+// Updates targeting keys with deal prefixes if minimum deal tier exceeded
+func applyDealSupport(bidRequest *openrtb.BidRequest, auc *auction) []error {
```
Comment (SyntaxNode) [inline exchange/exchange.go]: Please update comment to follow Go conventions:
```
applyDealSupport updates targeting keys....
```

File: exchange/exchange.go [hunk 12]
```diff
@@ -167,12 +168,93 @@ func (e *exchange) HoldAuction(ctx context.Context, bidRequest *openrtb.BidReque
...
+		}
+
+		impDealMap[imp.ID] = &bidderDealTier
+	}
+
+	return impDealMap
+}
+
+func validateDealTier(impDeal *DealTier) bool {
+	if impDeal.Info == nil {
```
Comment (SyntaxNode) [inline exchange/exchange.go]: Please add check for `impDeal == nil`. If it's not found in the DealInfo map, the impDeal will be nil here.

File: exchange/exchange.go [hunk 13]
```diff
@@ -167,12 +168,93 @@ func (e *exchange) HoldAuction(ctx context.Context, bidRequest *openrtb.BidReque
...
+			continue
+		}
+
+		impDealMap[imp.ID] = &bidderDealTier
+	}
+
+	return impDealMap
+}
+
+func validateDealTier(impDeal *DealTier) bool {
```
Comment (SyntaxNode) [inline exchange/exchange.go]: Please rename to `validateAndNormalizeDealTier` or `validateAndCleanDealTier` to make the mutation on line 240 obvious.

File: exchange/exchange.go (hunk 11 above)
Comment (camrice) [reply exchange/exchange.go]: Updated comment to follow conventions

File: exchange/exchange.go (hunk 12 above)
Comment (camrice) [reply exchange/exchange.go]: Added check for this case

File: exchange/exchange.go (hunk 13 above)
Comment (camrice) [reply exchange/exchange.go]: Renamed to `validateAndNormalizeDealTier` to make usage more clear

**Comment by ShriprasadM** [comment]
@SyntaxNode and @hhhjort : As per prebid documentation http://prebid.org/prebid-video/video-long-form.html#deal-support, `prioritizeDeals` parameter is used to convey if deals should be given a higher preference. But in this case, the parameter name we are expecting is `SupportDeals`. Can't we make it same as of Prebid i.e. `prioritizeDeals`? It will keep the nomenclature consistent.
